### ISSUE TITLE
kitchen test with serverspec fails: `run_ruby_script!': undefined method `success?' for nil:NilClass (NoMethodError)

### ISSUE THREAD

**Issue by tknerr** [test-kitchen/busser-serverspec#22]
Ohai!

I'm having a weird behaviour where the `kitchen test` reproducibly fails with the below error:
```
C:\Repos\_github\bills-kitchen\target\build\repo\vagrant-workflow-tests\playground\sample-toplevel-cookbook>bundle exec kitchen test
...
<snip>
...
       Suite path directory /tmp/busser/suites does not exist, skipping.
Uploading /tmp/busser/suites/serverspec/localhost/default_spec.rb (mode=0644)
Uploading /tmp/busser/suites/serverspec/spec_helper.rb (mode=0644)
D      [SSH] vagrant@127.0.0.1:2201<{:user_known_hosts_file=>"/dev/null", :paranoid=>false, :keys_only=>true, :port=>"2201", :keys=>["C:/Repos/_github/bills-k
itchen/target/build/repo/vagrant-workflow-tests/playground/sample-toplevel-cookbook/.kitchen/kitchen-vagrant/default-ubuntu-1204/.vagrant/machines/default/vir
tualbox/private_key"], :user=>"vagrant"}> (sh -c 'BUSSER_ROOT="/tmp/busser" GEM_HOME="/tmp/busser/gems" GEM_PATH="/tmp/busser/gems" GEM_CACHE="/tmp/busser/gem
s/cache" ; export BUSSER_ROOT GEM_HOME GEM_PATH GEM_CACHE; sudo -E /tmp/busser/bin/busser test')
-----> Running serverspec test suite
-----> Installing Serverspec..
Fetching: rspec-support-3.1.2.gem (100%)
Fetching: rspec-core-3.1.7.gem (100%)
Fetching: diff-lcs-1.2.5.gem (100%)
Fetching: rspec-expectations-3.1.2.gem (100%)
Fetching: rspec-mocks-3.1.3.gem (100%)
Fetching: rspec-3.1.0.gem (100%)
Fetching: rspec-its-1.1.0.gem (100%)
Fetching: multi_json-1.10.1.gem (100%)
Fetching: net-ssh-2.9.1.gem (100%)
Fetching: net-scp-1.2.1.gem (100%)
Fetching: specinfra-2.11.2.gem (100%)
Fetching: serverspec-2.7.1.gem (100%)
-----> serverspec installed (version 2.7.1)
/tmp/busser/gems/gems/busser-0.6.2/lib/busser/ui.rb:58:in `run_ruby_script!': undefined method `success?' for nil:NilClass (NoMethodError)
        from /tmp/busser/gems/gems/busser-serverspec-0.5.3/lib/busser/runner_plugin/serverspec.rb:36:in `test'
        from /tmp/busser/gems/gems/thor-0.19.0/lib/thor/command.rb:27:in `run'
        from /tmp/busser/gems/gems/thor-0.19.0/lib/thor/invocation.rb:126:in `invoke_command'
        from /tmp/busser/gems/gems/thor-0.19.0/lib/thor/invocation.rb:133:in `block in invoke_all'
        from /tmp/busser/gems/gems/thor-0.19.0/lib/thor/invocation.rb:133:in `each'
        from /tmp/busser/gems/gems/thor-0.19.0/lib/thor/invocation.rb:133:in `map'
        from /tmp/busser/gems/gems/thor-0.19.0/lib/thor/invocation.rb:133:in `invoke_all'
        from /tmp/busser/gems/gems/thor-0.19.0/lib/thor/group.rb:232:in `dispatch'
        from /tmp/busser/gems/gems/thor-0.19.0/lib/thor/invocation.rb:115:in `invoke'
        from /tmp/busser/gems/gems/busser-0.6.2/lib/busser/command/test.rb:43:in `block in perform'
        from /tmp/busser/gems/gems/busser-0.6.2/lib/busser/command/test.rb:35:in `each'
        from /tmp/busser/gems/gems/busser-0.6.2/lib/busser/command/test.rb:35:in `perform'
        from /tmp/busser/gems/gems/thor-0.19.0/lib/thor/command.rb:27:in `run'
        from /tmp/busser/gems/gems/thor-0.19.0/lib/thor/invocation.rb:126:in `invoke_command'
        from /tmp/busser/gems/gems/thor-0.19.0/lib/thor/invocation.rb:133:in `block in invoke_all'
        from /tmp/busser/gems/gems/thor-0.19.0/lib/thor/invocation.rb:133:in `each'
        from /tmp/busser/gems/gems/thor-0.19.0/lib/thor/invocation.rb:133:in `map'
        from /tmp/busser/gems/gems/thor-0.19.0/lib/thor/invocation.rb:133:in `invoke_all'
        from /tmp/busser/gems/gems/thor-0.19.0/lib/thor/group.rb:232:in `dispatch'
        from /tmp/busser/gems/gems/thor-0.19.0/lib/thor/invocation.rb:115:in `invoke'
        from /tmp/busser/gems/gems/thor-0.19.0/lib/thor.rb:40:in `block in register'
        from /tmp/busser/gems/gems/thor-0.19.0/lib/thor/command.rb:27:in `run'
        from /tmp/busser/gems/gems/thor-0.19.0/lib/thor/invocation.rb:126:in `invoke_command'
        from /tmp/busser/gems/gems/thor-0.19.0/lib/thor.rb:359:in `dispatch'
        from /tmp/busser/gems/gems/thor-0.19.0/lib/thor/base.rb:440:in `start'
        from /tmp/busser/gems/gems/busser-0.6.2/bin/busser:8:in `<top (required)>'
        from /tmp/busser/gems/bin/busser:23:in `load'
        from /tmp/busser/gems/bin/busser:23:in `<main>'
>>>>>> Verify failed on instance <default-ubuntu-1204>.
>>>>>> Please see .kitchen/logs/default-ubuntu-1204.log for more details
>>>>>> ------Exception-------
>>>>>> Class: Kitchen::ActionFailed
>>>>>> Message: SSH exited (1) for command: [sh -c 'BUSSER_ROOT="/tmp/busser" GEM_HOME="/tmp/busser/gems" GEM_PATH="/tmp/busser/gems" GEM_CACHE="/tmp/busser/g
ems/cache" ; export BUSSER_ROOT GEM_HOME GEM_PATH GEM_CACHE; sudo -E /tmp/busser/bin/busser test']
>>>>>> ----------------------
D      Verify failed on instance <default-ubuntu-1204>.
D      ------Exception-------
D      Class: Kitchen::InstanceFailure
D      Message: Verify failed on instance <default-ubuntu-1204>.  Please see .kitchen/logs/default-ubuntu-1204.log for more details
D      ---Nested Exception---
D      Class: Kitchen::ActionFailed
D      Message: SSH exited (1) for command: [sh -c 'BUSSER_ROOT="/tmp/busser" GEM_HOME="/tmp/busser/gems" GEM_PATH="/tmp/busser/gems" GEM_CACHE="/tmp/busser/gems/cache" ; export BUSSER_ROOT GEM_HOME GEM_PATH GEM_CACHE; sudo -E /tmp/busser/bin/busser test']
D      ------Backtrace-------
D      C:/Repos/_github/bills-kitchen/target/build/tools/chefdk/embedded/lib/ruby/gems/2.0.0/gems/test-kitchen-1.2.1/lib/kitchen/driver/ssh_base.rb:112:in `rescue in run_remote'
D      C:/Repos/_github/bills-kitchen/target/build/tools/chefdk/embedded/lib/ruby/gems/2.0.0/gems/test-kitchen-1.2.1/lib/kitchen/driver/ssh_base.rb:108:in `run_remote'
D      C:/Repos/_github/bills-kitchen/target/build/tools/chefdk/embedded/lib/ruby/gems/2.0.0/gems/test-kitchen-1.2.1/lib/kitchen/driver/ssh_base.rb:63:in `block in verify'
D      C:/Repos/_github/bills-kitchen/target/build/tools/chefdk/embedded/lib/ruby/gems/2.0.0/gems/test-kitchen-1.2.1/lib/kitchen/ssh.rb:47:in `initialize'
D      C:/Repos/_github/bills-kitchen/target/build/tools/chefdk/embedded/lib/ruby/gems/2.0.0/gems/test-kitchen-1.2.1/lib/kitchen/driver/ssh_base.rb:61:in `new'
D      C:/Repos/_github/bills-kitchen/target/build/tools/chefdk/embedded/lib/ruby/gems/2.0.0/gems/test-kitchen-1.2.1/lib/kitchen/driver/ssh_base.rb:61:in `verify'
D      C:/Repos/_github/bills-kitchen/target/build/tools/chefdk/embedded/lib/ruby/gems/2.0.0/gems/kitchen-vagrant-0.15.0/lib/kitchen/driver/vagrant.rb:85:in `verify'
D      C:/Repos/_github/bills-kitchen/target/build/tools/chefdk/embedded/lib/ruby/gems/2.0.0/gems/test-kitchen-1.2.1/lib/kitchen/instance.rb:273:in `public_send'
D      C:/Repos/_github/bills-kitchen/target/build/tools/chefdk/embedded/lib/ruby/gems/2.0.0/gems/test-kitchen-1.2.1/lib/kitchen/instance.rb:273:in `block in perform_action'
D      C:/Repos/_github/bills-kitchen/target/build/tools/chefdk/embedded/lib/ruby/gems/2.0.0/gems/test-kitchen-1.2.1/lib/kitchen/instance.rb:308:in `call'
D      C:/Repos/_github/bills-kitchen/target/build/tools/chefdk/embedded/lib/ruby/gems/2.0.0/gems/test-kitchen-1.2.1/lib/kitchen/instance.rb:308:in `synchronize_or_call'
D      C:/Repos/_github/bills-kitchen/target/build/tools/chefdk/embedded/lib/ruby/gems/2.0.0/gems/test-kitchen-1.2.1/lib/kitchen/instance.rb:283:in `block in action'
D      C:/Repos/_github/bills-kitchen/target/build/tools/chefdk/embedded/lib/ruby/2.0.0/benchmark.rb:281:in `measure'
D      C:/Repos/_github/bills-kitchen/target/build/tools/chefdk/embedded/lib/ruby/gems/2.0.0/gems/test-kitchen-1.2.1/lib/kitchen/instance.rb:282:in `action'
D      C:/Repos/_github/bills-kitchen/target/build/tools/chefdk/embedded/lib/ruby/gems/2.0.0/gems/test-kitchen-1.2.1/lib/kitchen/instance.rb:273:in `perform_action'
D      C:/Repos/_github/bills-kitchen/target/build/tools/chefdk/embedded/lib/ruby/gems/2.0.0/gems/test-kitchen-1.2.1/lib/kitchen/instance.rb:264:in `verify_action'
D      C:/Repos/_github/bills-kitchen/target/build/tools/chefdk/embedded/lib/ruby/gems/2.0.0/gems/test-kitchen-1.2.1/lib/kitchen/instance.rb:246:in `block in transition_to'                                                                                                                                                               
D      C:/Repos/_github/bills-kitchen/target/build/tools/chefdk/embedded/lib/ruby/gems/2.0.0/gems/test-kitchen-1.2.1/lib/kitchen/instance.rb:245:in `each'                   
D      C:/Repos/_github/bills-kitchen/target/build/tools/chefdk/embedded/lib/ruby/gems/2.0.0/gems/test-kitchen-1.2.1/lib/kitchen/instance.rb:245:in `transition_to'                                                                                                                                                                        
D      C:/Repos/_github/bills-kitchen/target/build/tools/chefdk/embedded/lib/ruby/gems/2.0.0/gems/test-kitchen-1.2.1/lib/kitchen/instance.rb:141:in `verify'                 
D      C:/Repos/_github/bills-kitchen/target/build/tools/chefdk/embedded/lib/ruby/gems/2.0.0/gems/test-kitchen-1.2.1/lib/kitchen/instance.rb:170:in `block in test'                                                                                                                                                                        
D      C:/Repos/_github/bills-kitchen/target/build/tools/chefdk/embedded/lib/ruby/2.0.0/benchmark.rb:281:in `measure'                                                        
D      C:/Repos/_github/bills-kitchen/target/build/tools/chefdk/embedded/lib/ruby/gems/2.0.0/gems/test-kitchen-1.2.1/lib/kitchen/instance.rb:166:in `test'                   
D      C:/Repos/_github/bills-kitchen/target/build/tools/chefdk/embedded/lib/ruby/gems/2.0.0/gems/test-kitchen-1.2.1/lib/kitchen/command.rb:109:in `public_send'                                                                                                                                                                           
D      C:/Repos/_github/bills-kitchen/target/build/tools/chefdk/embedded/lib/ruby/gems/2.0.0/gems/test-kitchen-1.2.1/lib/kitchen/command.rb:109:in `block (2 levels) in run_action'                                                                                                                                                        
D      ----------------------                                                                                                                                                
```

A subsequent `kitchen verify` works:
```
C:\Repos\_github\bills-kitchen\target\build\repo\vagrant-workflow-tests\playground\sample-toplevel-cookbook>bundle exec kitchen verify
-----> Starting Kitchen (v1.2.1)
-----> Verifying <default-ubuntu-1204>...
       Removing /tmp/busser/suites/serverspec
Uploading /tmp/busser/suites/serverspec/localhost/default_spec.rb (mode=0644)
Uploading /tmp/busser/suites/serverspec/spec_helper.rb (mode=0644)
-----> Running serverspec test suite
/opt/chef/embedded/bin/ruby -I/tmp/busser/suites/serverspec -I/tmp/busser/gems/gems/rspec-support-3.1.2/lib:/tmp/busser/gems/gems/rspec-core-3.1.7/lib /opt/chef/embedded/bin/rspec --pattern /tmp/busser/suites/serverspec/\*\*/\*_spec.rb --color --format documentation --default-path /tmp/busser/suites/serverspec

Package "apache2"
  should be installed

Service "apache2"
  should be enabled
  should be running

Port "80"
  should be listening

File "/var/www/sample.html"
  should be file
  should contain "I can haz cats!"

Finished in 0.11642 seconds (files took 0.29134 seconds to load)
6 examples, 0 failures
       Finished verifying <default-ubuntu-1204> (0m2.18s).
-----> Kitchen is finished. (0m5.43s)
```

Steps to reproduce:

* `git clone https://github.com/tknerr/sample-toplevel-cookbook.git`
* `cd sample-toplevel-cookbook`
* `git checkout 37ec35f7e3c43673d95eff556da455f12dcfdd1a`
* `bundle install`
* `rake integration` or `bundle exec kitchen test` directly

Any ideas what the problem could be?

**Comment by cl-lab-k** [comment]
reproduced.

**Comment by cl-lab-k** [comment]
@tknerr
can you forward this issue to test-kitchen/busser ?

**Comment by tknerr** [comment]
@cl-lab-k yes sure. I thought it would be more appropriate here by looking at the stacktrace only, but I hadn't looked any deeper in the code.

So here you go: test-kitchen/busser#17

**Comment by cl-lab-k** [comment]
i tested same testcase on busser-rspec using run_ruby_script! too:
https://github.com/test-kitchen/busser-rspec/blob/master/lib/busser/runner_plugin/rspec.rb#L54
but it ran successfully. so i do not know why crashes in special situations.

**Comment by tyler-ball** [comment]
@fnichol and I spent some time (too much) working on this.  The underlying issue (being concealed inside Thor) is:

```
/tmp/busser/gems/gems/thor-0.19.0/lib/thor/actions.rb:254:in ``': Cannot allocate memory - /opt/chef/embedded/bin/ruby /tmp/busser/gems/gems/busser-serverspec-0.5.3/lib/busser/runner_plugin/../serverspec/runner.rb /tmp/busser/suites/serverspec (Errno::ENOMEM)
```

I found this by modifying https://github.com/test-kitchen/busser/blob/master/lib/busser/ui.rb#L44 and turning `:capture` to true.

Because it is running out of memory I simply increased the VM memory by adding:
```yaml
driver:
  customize:
    memory: 2048
```

Can you try that?  It also seems like your .kitchen.yml is a little out of date according to the keys listed in https://github.com/test-kitchen/kitchen-vagrant.

Because of the ridiculous amount of time @fnichol and I spent debugging this, I'm going to recommend we turn `:capture` to true all the time as our part of this issue.  @fnichol do you know a reason we shouldn't do this?

**Comment by fnichol** [comment]
Hi all, there's an RC release of Busser (0.7.0.rc.1) on RubyGems with the fix in test-kitchen/busser#20. You can try it by adding the following to your Kitchen config:

```yaml
---
busser:
  version: busser@0.7.1.rc.1
```

Note that the `busser` block will be changing to a `verifier` block in Test Kitchen 1.4 (due out shortly), but the `busser` blocks will be support for backwards compat.

I'm looking to release Busser 0.7.0 later today, after which time everyone gets this update when they `kitchen verify`.

Thanks all!

**Comment by tknerr** [comment]
Thanks @tyler-ball and @fnichol for tracking this down!

@tyler-ball yes the .kitchen.yml was in an older format at this point in time, but I switched to the newer format one commit after or so. As for the vagrant keys specifically, lots has changed there in master and I'm hoping for a new release soon, now that @fnichol is working on it again :-)

Let me check if I can still reproduce the issue...

**Comment by tknerr** [comment]
@tyler-ball @fnichol could not reproduce this anymore, or to say it the other way around: can verify that it is fixed now that busser 0.7.0 is out and released! :-)

Thanks guys! :beers:
